src/librustc_lint/builtin.rs

```diff
@@ -398,43 +396,256 @@ struct ImproperCTypesVisitor<'a, 'tcx: 'a> {
+                for variant in variants {
+                    for arg in &variant.args {
+                        let arg = normalize_type(cx, arg);
+                        let r = self.check_type_for_ffi(cache, arg);
+                        match r {
+                            FfiSafe => {}
+                            FfiBadStruct(..) | FfiBadEnum(..) => { return r; }
+                            FfiUnsafe(s) => { return FfiBadEnum(did, s); }
+                        }
+                    }
+                }
+                FfiSafe
+            }
+            ty::TyBox(..) => {
+                FfiUnsafe("found Rust type Box<_> in foreign module, \
+                           consider using a raw pointer instead")
+            }
+            ty::TySlice(_) => {
+                FfiUnsafe("found Rust slice type in foreign module, \
+                           consider using a raw pointer instead")
+            }
+            ty::TyTrait(..) => {
+                FfiUnsafe("found Rust trait type in foreign module, \
+                           consider using a raw pointer instead")
+            }
+            ty::TyInt(ast::TyIs) => {
+                FfiUnsafe("found Rust type `isize` in foreign module, while \
+                          `libc::c_int` or `libc::c_long` should be used")
+            }
+            ty::TyUint(ast::TyUs) => {
+                FfiUnsafe("found Rust type `usize` in foreign module, while \
+                          `libc::c_uint` or `libc::c_ulong` should be used")
+            }
+            ty::TyChar => {
+                FfiUnsafe("found Rust type `char` in foreign module, while \
+                           `u32` or `libc::wchar_t` should be used")
+            }
+            ty::TyStr => {
+                FfiUnsafe("found Rust type `str` in foreign module; \
+                           consider using a `*const libc::c_char`")
+            }
+            ty::TyTuple(ref tys) => {
+                // The layout of tuples is implicitly repr(C).
+                // FIXME: Is that correct?
+
+                // Check the fields of the tuple.
+                for field in tys {
+                    let r = self.check_type_for_ffi(cache, field);
+                    match r {
+                        FfiSafe => {}
+                        _ => return r
+                    }
+                }
+                FfiSafe
+            }
+
+            ty::TyRawPtr(ref m) | ty::TyRef(_, ref m) => {
+                self.check_type_for_ffi(cache, m.ty)
+            }
+
+            // FIXME: Maybe we want to prohibit arrays of zero-size types?
+            ty::TyArray(ty, _) => {
+                self.check_type_for_ffi(cache, ty)
+            }
+
+            ty::TyBareFn(..) => {
+                // FIXME: Implement
+                FfiSafe
+            }
+
+            // Primitive types with a stable representation.
+            ty::TyBool | ty::TyInt(..) | ty::TyUint(..) |
```

Comment:
If you were writing a malloc replacement in Rust, for example, you'd likely use size/usize internally, and it seems likely that you'd want  to use those as well in your extern signature for tjmalloc or whatever. 

The alternative is to use libc's size_t type alias in those signatures, but that seems to just be shoving the problem elsewhere.

> On Jul 21, 2015, at 14:38, Nick Cameron <notifications@github.com> wrote:
> 
> In src/librustc_lint/builtin.rs:
> 
> > +            ty::TyRawPtr(ref m) | ty::TyRef(_, ref m) => {
> > +                self.check_type_for_ffi(cache, m.ty)
> > +            }
> > +
> > +            // FIXME: Maybe we want to prohibit arrays of zero-size types?
> > +            ty::TyArray(ty, _) => {
> > +                self.check_type_for_ffi(cache, ty)
> > +            }
> > +
> > +            ty::TyBareFn(..) => {
> > +                // FIXME: Implement
> > +                FfiSafe
> > +            }
> > +
> > +            // Primitive types with a stable representation.
> > +            ty::TyBool | ty::TyInt(..) | ty::TyUint(..) |
> Shouldn't usize and isize be FfiUnsafe?
> 
> —
> Reply to this email directly or view it on GitHub.
>